hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java

```diff
@@ -7369,6 +7367,38 @@ public Object performTask() throws IOException {
     }.handle(this);
   }
 
+  @VisibleForTesting
+  void unlockSubtree(final String path, final long ignoreStoInodeId) throws IOException {
+    boolean failed = true;
+    try{
+      unlockSubtreeInternal(path, ignoreStoInodeId);
+    }catch(Exception e ){
+      //Unable to remove the lock. setting the async removal flag
+      setAsyncLockRemoval(path);
+      throw e;
+    }
+  }
+
+  void setAsyncLockRemoval(final String path) throws IOException {
+    LightWeightRequestHandler handler = new LightWeightRequestHandler(
+            HDFSOperationType.SET_ASYNC_SUBTREE_RECOVERY_FLAG) {
+      @Override
+      public Object performTask() throws IOException {
+        OngoingSubTreeOpsDataAccess<SubTreeOperation> dataAccess =
+                (OngoingSubTreeOpsDataAccess) HdfsStorageFactory.getDataAccess(OngoingSubTreeOpsDataAccess.class);
+        SubTreeOperation op = dataAccess.findByPath(getSubTreeLockPathPrefix(path));
+        if (op != null) {
+          op.setAsyncLockRecoveryTime(System.currentTimeMillis());
```

Comment:
setAsyncLockRemoval function is called only once.  it is not possible to set this flag multiple times.

Reply:
From what I understand setAsyncLockRemoval is called each time unlockSubtree is called and unlockSubtreeInternal fails.
unlockSubtree is also called in the cleaner so if it fails in the cleaner it will overwrite the date from the original failure and replace it by the most recent cleaner run, it would be good to have the date of the original failure instead.

Reply:
fixed